nameko/testing/utils.py

```diff
@@ -140,3 +140,65 @@ def find_free_port(host='127.0.0.1'):
     port = sock.getsockname()[1]
     sock.close()
     return port
+
+
+class ResourcePipeline(object):
+    """ Creates and destroys resources in background threads.
+
+    Creates up to `size` resources ahead of time so the caller avoids waiting
+    for lazy creation.
+    """
+
+    STOP = object()
+
+    def __init__(self, create, destroy, size=3):
+        if size == 0:
+            raise RuntimeError("Zero size would create unbounded resources")
+        self.ready = eventlet.Queue(maxsize=size)
+        self.trash = eventlet.Queue()
+
+        self.threads = []
+        self.create = create
+        self.destroy = destroy
+
+    def start(self):
+        self.running = True
+        self.threads.append(eventlet.spawn(self._create))
+        self.threads.append(eventlet.spawn(self._destroy))
+
+    def _create(self):
+        while self.running:
+            obj = self.create()
+            self.ready.put(obj)
+
+    def _destroy(self):
+        while True:
+            obj = self.trash.get()
+            if obj is ResourcePipeline.STOP:
+                break
+            self.destroy(obj)
+
+    def get(self):
+        return self.ready.get()
+
+    def discard(self, vhost):
+        self.trash.put(vhost)
+
+    def shutdown(self):
+        self.running = False
+        while self.ready.qsize():
+            unused = self.ready.get()
+            self.trash.put(unused)
+            eventlet.sleep()  # allow create thread to run
```

Comment:
It's for the case where the create loop is blocked waiting to put an item into the `ready` queue (i.e. waiting for it to be less than maxsize).

`unused = self.ready.get()` unblocks the create thread but doesn't yield, so we exit the while loop here before giving it a chance to run. Once unblocked, the create thread deposits its item into the ready queue, and if we've already exited the loop, that final item isn't destroyed.

I've expanded the inline comment to hopefully explain